python/ray/data/tests/test_dataset_tfrecords.py

```diff
@@ -177,28 +218,38 @@ def test_readback_tfrecords(ray_start_regular_shared, tmp_path):
             {
                 "int_item": 1,
                 "int_list": [2, 2, 3],
+                "int_empty": [],
                 "float_item": 1.0,
                 "float_list": [2.0, 3.0, 4.0],
+                "float_empty": 1.0,
                 "bytes_item": b"abc",
-                "bytes_list": [b"abc", b"1234"],
+                "bytes_list": [b"def", b"1234"],
+                "bytes_empty": None,
             },
             # Row two.
             {
                 "int_item": 2,
                 "int_list": [3, 3, 4],
+                "int_empty": [9, 2],
                 "float_item": 2.0,
-                "float_list": [2.0, 2.0, 3.0],
-                "bytes_item": b"def",
-                "bytes_list": [b"def", b"1234"],
+                "float_list": [5.0, 6.0, 7.0],
+                "float_empty": None,
+                "bytes_item": b"ghi",
+                "bytes_list": [b"jkl", b"5678"],
+                "bytes_empty": b"hello",
             },
-        ]
+        ],
+        # Here and in the read_tfrecords call below, we specify `parallelism=1`
+        # to ensure that all rows end up in the same block, which is required
+        # for type inference involving partially missing columns.
+        parallelism=1,
     )
 
     # Write the TFRecords.
     ds.write_tfrecords(tmp_path)
 
     # Read the TFRecords.
-    readback_ds = ray.data.read_tfrecords(tmp_path)
+    readback_ds = ray.data.read_tfrecords(tmp_path, parallelism=1)
```

Comment:
Could we actually remove this `parallelism` specification, along with the above `parallelism` specification for `from_items()`, and change the below assert to:
```python
# Repartition each dataset into single block so column types are unified.
assert ds.repartition(1).take() == readback_ds.repartition(1).take()
```
That way we can ensure that multi-block roundtrips work (provided that it does actually work, lmk if it doesn't), and the single-block thing is only part of checking that the roundtrip is equal.